<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CKMRsim-example-2-microsatellites • CKMRsim</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CKMRsim-example-2-microsatellites">
<meta property="og:description" content="CKMRsim">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CKMRsim</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CKMRsim-developer-notes.html">CKMRsim-developer-notes</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-example-1.html">CKMRsim-example-1</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-example-2-microsatellites.html">CKMRsim-example-2-microsatellites</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-pairwise-relationships.html">CKMRsim-pairwise-relationships</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-simulating-linked-markers.html">Simulating-linked-markers</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-writing-geno-error-funcs.html">CKMRsim-writing-geno-error-functions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="CKMRsim-example-2-microsatellites_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CKMRsim-example-2-microsatellites</h1>
                        <h4 class="author">Eric C. Anderson</h4>
            
            <h4 class="date">2021-02-01</h4>
      
      
      <div class="hidden name"><code>CKMRsim-example-2-microsatellites.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Now that y’all have, presumably, worked your way through the CKMRsim-example-1 vignette, we will run through another situation here (more quickly, and with considerably less discussion) in a study using next-generation-sequencing genotyped microsatellites on Atlantic salmon (<em>Salmo salar</em>).</p>
<p>Before we get going, let’s load some libraries:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">CKMRsim</span><span class="op">)</span></code></pre></div>
<div id="data-set" class="section level3">
<h3 class="hasAnchor">
<a href="#data-set" class="anchor"></a>Data Set</h3>
<p>We start with the data for all of the populations from <span class="citation">Bradbury et al. (2018)</span>. These data come with the CKMRsim package in a wide-format tibble similar to the format used for the package <code>rubias</code> (minus a few columns).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">labrador_salar_ngs_msats</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt; # A tibble: 5 x 5</span>
<span class="co">#&gt;   collection indiv                  `NGS-SsaD486` `NGS-SsaD486_1` `NGS-SSsp2210`</span>
<span class="co">#&gt;   &lt;chr&gt;      &lt;chr&gt;                  &lt;chr&gt;         &lt;chr&gt;           &lt;chr&gt;         </span>
<span class="co">#&gt; 1 ALRPARR    ALRPARR_ALRPARR_003702 076           080             032           </span>
<span class="co">#&gt; 2 ALRPARR    ALRPARR_ALRPARR_003705 080           080             052           </span>
<span class="co">#&gt; 3 ALRPARR    ALRPARR_ALRPARR_003706 &lt;NA&gt;          &lt;NA&gt;            &lt;NA&gt;          </span>
<span class="co">#&gt; 4 ALRPARR    ALRPARR_ALRPARR_003707 076           076             044           </span>
<span class="co">#&gt; 5 ALRPARR    ALRPARR_ALRPARR_003709 076           080             052</span></code></pre></div>
<p>We will develop a few analysis steps working on a single collection (POTMIDBTLUWST) from this data set. We will make a point of writing a function for each step, so that, when we are done, we might try repeating the whole affair for every collection in the data set, getting our tidyverse/purrr on for that job.</p>
<p>Here we grab just that one collection:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">D</span> <span class="op">&lt;-</span> <span class="va">labrador_salar_ngs_msats</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">collection</span> <span class="op">==</span> <span class="st">"POTMIDBTLUWST"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">collection</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="preparing-data-for-use-with-ckmrsim" class="section level2">
<h2 class="hasAnchor">
<a href="#preparing-data-for-use-with-ckmrsim" class="anchor"></a>Preparing data for use with CKMRsim</h2>
<p>First, make a long-format data frame like CKMR expects for genotypes:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">make_long</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">D</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">nc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">D</span><span class="op">)</span>
  <span class="va">loci</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">D</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">3</span>, <span class="va">nc</span>, by <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">]</span>
  <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">D</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">3</span>, <span class="va">nc</span>, by <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">str_c</span><span class="op">(</span><span class="va">loci</span>, <span class="st">"---"</span>, <span class="st">"2"</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">D</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">2</span>, <span class="va">nc</span>, by <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">str_c</span><span class="op">(</span><span class="va">loci</span>, <span class="st">"---"</span>, <span class="st">"1"</span><span class="op">)</span>
  
  <span class="va">D</span> <span class="op">%&gt;%</span> 
    <span class="fu">gather</span><span class="op">(</span>key <span class="op">=</span> <span class="st">"loccopy"</span>, value <span class="op">=</span> <span class="st">"Allele"</span>, <span class="op">-</span><span class="va">indiv</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">separate</span><span class="op">(</span><span class="va">loccopy</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Locus"</span>, <span class="st">"gene_copy"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"---"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">rename</span><span class="op">(</span>Indiv <span class="op">=</span> <span class="va">indiv</span><span class="op">)</span>
<span class="op">}</span>

<span class="va">L</span> <span class="op">&lt;-</span> <span class="fu">make_long</span><span class="op">(</span><span class="va">D</span><span class="op">)</span></code></pre></div>
<p>And once we have that we need to summarise the allele frequencies. I really should have a CKMR function for this. Note that we will drop monomorphic loci here.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">long2freqs</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">L</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">L</span> <span class="op">%&gt;%</span>
    <span class="fu">group_by</span><span class="op">(</span><span class="va">Locus</span>, <span class="va">Allele</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">tally</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">Allele</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>Freq <span class="op">=</span> <span class="va">n</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">arrange</span><span class="op">(</span><span class="va">Locus</span>, <span class="fu">desc</span><span class="op">(</span><span class="va">Freq</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">n</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">group_by</span><span class="op">(</span><span class="va">Locus</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># this removes monomorphic loci</span>
    <span class="fu">mutate</span><span class="op">(</span>AlleIdx <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">mutate</span><span class="op">(</span>LocIdx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">Locus</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">Locus</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,
           Chrom <span class="op">=</span> <span class="st">"Unk"</span>,  <span class="co"># Ian probably knows where these are, but I don't have that info at the moment.</span>
           Pos <span class="op">=</span> <span class="va">LocIdx</span>
           <span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu">select</span><span class="op">(</span><span class="va">Chrom</span>, <span class="va">Pos</span>, <span class="va">Locus</span>, <span class="va">Allele</span>, <span class="va">Freq</span>, <span class="va">LocIdx</span>, <span class="va">AlleIdx</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/reindex_markers.html">reindex_markers</a></span><span class="op">(</span><span class="op">)</span>  <span class="co"># this is a CKMRsim function</span>
<span class="op">}</span>

<span class="va">Fr</span> <span class="op">&lt;-</span> <span class="fu">long2freqs</span><span class="op">(</span><span class="va">L</span><span class="op">)</span></code></pre></div>
<p>Any loci that were monomorphic must also be dropped from the original genotype data set. So, we will include all of the above (and tossing monomorphic loci) in a function that returns two tibbles: 1) the long format genos for use (later) in CKMRsim, and 2) the allele freqs for use in CKMRsim.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prep_genos4CKMRsim</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">D</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">L</span> <span class="op">&lt;-</span> <span class="fu">make_long</span><span class="op">(</span><span class="va">D</span><span class="op">)</span>
  <span class="va">Fr</span> <span class="op">&lt;-</span> <span class="fu">long2freqs</span><span class="op">(</span><span class="va">L</span><span class="op">)</span>
  <span class="va">Lp</span> <span class="op">&lt;-</span> <span class="fu">semi_join</span><span class="op">(</span><span class="va">L</span>, <span class="va">Fr</span>, by <span class="op">=</span> <span class="st">"Locus"</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Genos <span class="op">=</span> <span class="va">Lp</span>, Freqs <span class="op">=</span> <span class="va">Fr</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># and now we do that all from the beginning again.</span>
<span class="co"># GaF = Genos and Freqs</span>
<span class="va">GaF</span> <span class="op">&lt;-</span> <span class="va">labrador_salar_ngs_msats</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">collection</span> <span class="op">==</span> <span class="st">"POTMIDBTLUWST"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">collection</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">prep_genos4CKMRsim</span><span class="op">(</span><span class="va">.</span><span class="op">)</span></code></pre></div>
</div>
<div id="estimating-power-for-relationship-inference" class="section level2">
<h2 class="hasAnchor">
<a href="#estimating-power-for-relationship-inference" class="anchor"></a>Estimating Power for Relationship Inference</h2>
<p>We will assume the microsatellite genotyping error model with a per-gene-copy miscall rate of 0.003 and a per-gene-copy average-length-allele dropout rate of 0.004, with the standard z-scale inflation factor. We will assume true and assumed genotyping error models are the same, for now…</p>
<p>We will do this all in terms of functions that will be easy to apply to all the collections later…</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ngs_msat_create_ckmr</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Fr</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="../reference/create_ckmr.html">create_ckmr</a></span><span class="op">(</span>D <span class="op">=</span> <span class="va">Fr</span>,
              kappa_matrix <span class="op">=</span> <span class="va">kappas</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PO"</span>, <span class="st">"FS"</span>, <span class="st">"HS"</span>, <span class="st">"U"</span><span class="op">)</span>, <span class="op">]</span>, 
              ge_mod_assumed <span class="op">=</span> <span class="va">ge_model_microsat1</span>, 
              ge_mod_true <span class="op">=</span> <span class="va">ge_model_microsat1</span>, 
              ge_mod_assumed_pars_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>miscall_rate <span class="op">=</span> <span class="fl">0.003</span>, dropout_rate <span class="op">=</span> <span class="fl">0.004</span><span class="op">)</span>, 
              ge_mod_true_pars_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>miscall_rate <span class="op">=</span> <span class="fl">0.003</span>, dropout_rate <span class="op">=</span> <span class="fl">0.004</span><span class="op">)</span>
            <span class="op">)</span>
<span class="op">}</span>

<span class="co"># and use it:</span>
<span class="va">CK</span> <span class="op">&lt;-</span> <span class="fu">ngs_msat_create_ckmr</span><span class="op">(</span><span class="va">GaF</span><span class="op">$</span><span class="va">Freqs</span><span class="op">)</span></code></pre></div>
<p>Then, let’s simulate 20,000 replicates. In production, you will probably want to do more. We will assume no missing data at this point, but that can be changed in a future run.d</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">simQs</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">CK</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="../reference/simulate_Qij.html">simulate_Qij</a></span><span class="op">(</span>C <span class="op">=</span> <span class="va">CK</span>, reps <span class="op">=</span> <span class="fl">2e04</span>,
              sim_relats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PO"</span>, <span class="st">"FS"</span>, <span class="st">"HS"</span>, <span class="st">"U"</span><span class="op">)</span>,
              calc_relats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PO"</span>, <span class="st">"FS"</span>, <span class="st">"U"</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>

<span class="va">Qs</span> <span class="op">&lt;-</span> <span class="fu">simQs</span><span class="op">(</span><span class="va">CK</span><span class="op">)</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span></code></pre></div>
<p>Now, let’s extract some logls (for PO and FS) so that we can plot them</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">logls</span> <span class="op">&lt;-</span> <span class="fu">bind_rows</span><span class="op">(</span>
  <span class="fu"><a href="../reference/extract_logls.html">extract_logls</a></span><span class="op">(</span><span class="va">Qs</span>, numer <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>PO <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, denom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>U <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span>,
  <span class="fu"><a href="../reference/extract_logls.html">extract_logls</a></span><span class="op">(</span><span class="va">Qs</span>, numer <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FS <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, denom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>U <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span>,
  <span class="fu"><a href="../reference/extract_logls.html">extract_logls</a></span><span class="op">(</span><span class="va">Qs</span>, numer <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>PO <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, denom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FS <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p>Now, we can plot those and get a picture of how we are doing with these killer microsats.</p>
<p>This plots the PO/U and FS/U log likelihood ratios.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">logls</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">denom_wts</span> <span class="op">==</span> <span class="st">"U=1"</span><span class="op">)</span> <span class="op">%&gt;%</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logl_ratio</span>, fill <span class="op">=</span> <span class="va">true_relat</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.35</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">numer_wts</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> </code></pre></div>
<p><img src="CKMRsim-example-2-microsatellites_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
<p>So, we see a lot of overlap between FS and PO, which we expect. Not much other overlap, so I suspect that FS and PO will easily be distinguished from unrelated.</p>
<p>How much overlap between FS and PO when we use the PO/FS statistic?</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">logls</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">denom_wts</span> <span class="op">==</span> <span class="st">"FS=1"</span>, <span class="va">numer_wts</span> <span class="op">==</span> <span class="st">"PO=1"</span><span class="op">)</span> <span class="op">%&gt;%</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logl_ratio</span>, fill <span class="op">=</span> <span class="va">true_relat</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.35</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">numer_wts</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> </code></pre></div>
<p><img src="CKMRsim-example-2-microsatellites_files/figure-html/unnamed-chunk-12-1.png" width="672"></p>
<p>Let’s now do importance sampling to see what the FPRs and FNRs are. We will get FPRs at a lot of FNRs spaced close together so that we can get the FNR that is closest to a “good” FPR (of about 1/100 the reciprocal of the number of comparisons). Let’s say for now that we are going to be shooting for 10e-10 (should be plenty low…).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mc_samp</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Q</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="../reference/mc_sample_simple.html">mc_sample_simple</a></span><span class="op">(</span>
    <span class="va">Q</span>, 
    nu <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PO"</span>, <span class="st">"FS"</span><span class="op">)</span>,
    FNRs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0.001</span>, <span class="fl">0.20</span>, by <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span><span class="op">)</span>  
  <span class="op">)</span>
<span class="op">}</span>

<span class="va">FPRs</span> <span class="op">&lt;-</span> <span class="fu">mc_samp</span><span class="op">(</span><span class="va">Qs</span><span class="op">)</span></code></pre></div>
<p>OK, now, we will want to choose the lowest FNR with an FPR &lt; 1e10 for both FS and PO, and return that in a tibble.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">get_low_FPR</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">R</span>, <span class="va">cutoff</span> <span class="op">=</span> <span class="fl">1e-10</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">R</span> <span class="op">%&gt;%</span>
    <span class="fu">group_by</span><span class="op">(</span><span class="va">numerator</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">FPR</span> <span class="op">&lt;</span> <span class="va">cutoff</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">slice</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="op">}</span>

<span class="va">low_FPR</span> <span class="op">&lt;-</span> <span class="fu">get_low_FPR</span><span class="op">(</span><span class="va">FPRs</span><span class="op">)</span>
<span class="va">low_FPR</span>
<span class="co">#&gt; # A tibble: 2 x 10</span>
<span class="co">#&gt; # Groups:   numerator [2]</span>
<span class="co">#&gt;     FNR      FPR       se num_nonzero_wts Lambda_star pstar mc_method numerator</span>
<span class="co">#&gt;   &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;           &lt;int&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;    </span>
<span class="co">#&gt; 1 0.061 9.56e-11 3.51e-12           18780        18.9 FS    IS        FS       </span>
<span class="co">#&gt; 2 0.003 8.72e-11 8.29e-12           19940        17.1 PO    IS        PO       </span>
<span class="co">#&gt; # … with 2 more variables: denominator &lt;chr&gt;, true_relat &lt;chr&gt;</span></code></pre></div>
<p>That is really encouraging.</p>
</div>
<div id="doing-relationship-inference" class="section level2">
<h2 class="hasAnchor">
<a href="#doing-relationship-inference" class="anchor"></a>Doing Relationship Inference</h2>
<p>Now, we can look for PO pairs and for FS pairs in the data set.</p>
<p>First, look for individuals that might be the same:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">matchers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_close_matching_genotypes.html">find_close_matching_genotypes</a></span><span class="op">(</span>
  LG <span class="op">=</span> <span class="va">GaF</span><span class="op">$</span><span class="va">Genos</span>, 
  CK <span class="op">=</span> <span class="va">CK</span>, 
  max_mismatch <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p>There is no one even close to matching…not even with 10 loci mismatching</p>
<p>Make a function to look for FS pairs (I think these guys were all sampled as juvies) and for each individual in the second data frame, just keep the top 10 hits.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">FS_candidates</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pairwise_kin_logl_ratios.html">pairwise_kin_logl_ratios</a></span><span class="op">(</span>
  D1 <span class="op">=</span> <span class="va">GaF</span><span class="op">$</span><span class="va">Genos</span>, 
  D2 <span class="op">=</span> <span class="va">GaF</span><span class="op">$</span><span class="va">Genos</span>, 
  CK <span class="op">=</span> <span class="va">CK</span>, 
  numer <span class="op">=</span> <span class="st">"FS"</span>, 
  denom <span class="op">=</span> <span class="st">"U"</span>, 
  keep_top <span class="op">=</span> <span class="fl">10</span>,
  num_cores <span class="op">=</span> <span class="fl">1</span>  <span class="co"># only on here for CRAN CHECK.  Remove for personal use</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">logl_ratio</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>

<span class="va">FS_candidates</span>
<span class="co">#&gt; # A tibble: 615 x 4</span>
<span class="co">#&gt;    D2_indiv                      D1_indiv                     logl_ratio num_loc</span>
<span class="co">#&gt;    &lt;chr&gt;                         &lt;chr&gt;                             &lt;dbl&gt;   &lt;int&gt;</span>
<span class="co">#&gt;  1 POTMIDBTLUWST_POTMIDBTLUWST_… POTMIDBTLUWST_POTMIDBTLUWST…      26.4       92</span>
<span class="co">#&gt;  2 POTMIDBTLUWST_POTMIDBTLUWST_… POTMIDBTLUWST_POTMIDBTLUWST…      20.4       94</span>
<span class="co">#&gt;  3 POTMIDBTLUWST_POTMIDBTLUWST_… POTMIDBTLUWST_POTMIDBTLUWST…      15.2       94</span>
<span class="co">#&gt;  4 POTMIDBTLUWST_POTMIDBTLUWST_… POTMIDBTLUWST_POTMIDBTLUWST…      14.1       91</span>
<span class="co">#&gt;  5 POTMIDBTLUWST_POTMIDBTLUWST_… POTMIDBTLUWST_POTMIDBTLUWST…      13.1       75</span>
<span class="co">#&gt;  6 POTMIDBTLUWST_POTMIDBTLUWST_… POTMIDBTLUWST_POTMIDBTLUWST…      13.1       64</span>
<span class="co">#&gt;  7 POTMIDBTLUWST_POTMIDBTLUWST_… POTMIDBTLUWST_POTMIDBTLUWST…      12.1       91</span>
<span class="co">#&gt;  8 POTMIDBTLUWST_POTMIDBTLUWST_… POTMIDBTLUWST_POTMIDBTLUWST…      10.1       83</span>
<span class="co">#&gt;  9 POTMIDBTLUWST_POTMIDBTLUWST_… POTMIDBTLUWST_POTMIDBTLUWST…       9.80      58</span>
<span class="co">#&gt; 10 POTMIDBTLUWST_POTMIDBTLUWST_… POTMIDBTLUWST_POTMIDBTLUWST…       8.05      89</span>
<span class="co">#&gt; # … with 605 more rows</span></code></pre></div>
<p>OK, that is pretty cool.</p>
<div id="do-all-the-pairs-and-plot-their-distribution" class="section level3">
<h3 class="hasAnchor">
<a href="#do-all-the-pairs-and-plot-their-distribution" class="anchor"></a>Do all the pairs and plot their distribution</h3>
<p>As Mark said, it is wortwhile to compare the distribution of logl_ratios for (unrelated) individuals to the simulated values to make sure there is nothing super weird going on. Let’s do that.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">all_FS_comparisons</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pairwise_kin_logl_ratios.html">pairwise_kin_logl_ratios</a></span><span class="op">(</span>
  D1 <span class="op">=</span> <span class="va">GaF</span><span class="op">$</span><span class="va">Genos</span>, 
  D2 <span class="op">=</span> <span class="va">GaF</span><span class="op">$</span><span class="va">Genos</span>, 
  CK <span class="op">=</span> <span class="va">CK</span>, 
  numer <span class="op">=</span> <span class="st">"FS"</span>, 
  denom <span class="op">=</span> <span class="st">"U"</span>,
  num_cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span></code></pre></div>
<p>And plot.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DD</span> <span class="op">&lt;-</span> <span class="va">logls</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">true_relat</span> <span class="op">==</span> <span class="st">"U"</span>, <span class="va">numer_wts</span> <span class="op">==</span> <span class="st">"FS=1"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="va">logl_ratio</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>what <span class="op">=</span> <span class="st">"simulated"</span><span class="op">)</span>

<span class="va">to_plot</span> <span class="op">&lt;-</span> <span class="va">all_FS_comparisons</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="va">logl_ratio</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>what <span class="op">=</span> <span class="st">"observed"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">bind_rows</span><span class="op">(</span><span class="va">.</span>, <span class="va">DD</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">to_plot</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logl_ratio</span>, fill <span class="op">=</span> <span class="va">what</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></code></pre></div>
<p><img src="CKMRsim-example-2-microsatellites_files/figure-html/unnamed-chunk-18-1.png" width="672"></p>
<p>Aha! That is likely due to the population structure (this was a mixture of collections from several different places.)</p>
</div>
</div>
<div id="now-lets-do-that-with-every-collection" class="section level2">
<h2 class="hasAnchor">
<a href="#now-lets-do-that-with-every-collection" class="anchor"></a>Now, let’s do that with every collection…</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">first_steps</span> <span class="op">&lt;-</span> <span class="va">labrador_salar_ngs_msats</span> <span class="op">%&gt;%</span> 
  <span class="fu">group_by</span><span class="op">(</span><span class="va">collection</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">nest</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="co"># make data for each collection a tibble in a list column</span>
  <span class="fu">mutate</span><span class="op">(</span>GaF <span class="op">=</span> <span class="fu">map</span><span class="op">(</span>.x <span class="op">=</span> <span class="va">data</span>, .f <span class="op">=</span> <span class="va">prep_genos4CKMRsim</span><span class="op">)</span>, <span class="co"># prep geons and alle freqs</span>
         CK <span class="op">=</span> <span class="fu">map</span><span class="op">(</span><span class="va">GaF</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">ngs_msat_create_ckmr</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">Freqs</span><span class="op">)</span><span class="op">)</span>, <span class="co"># make ckmr objects</span>
         Qs <span class="op">=</span> <span class="fu">map</span><span class="op">(</span><span class="va">CK</span>, <span class="va">simQs</span><span class="op">)</span>
         <span class="op">)</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: HS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span></code></pre></div>
<p>Now, we can sample those and get some FPRs, etc.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">next_step</span> <span class="op">&lt;-</span> <span class="va">first_steps</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>FPRs <span class="op">=</span> <span class="fu">map</span><span class="op">(</span><span class="va">Qs</span>, <span class="va">mc_samp</span><span class="op">)</span>,
         lowFPR <span class="op">=</span> <span class="fu">map</span><span class="op">(</span><span class="va">FPRs</span>, <span class="va">get_low_FPR</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>From that we see that a few of the populations don’t have power for relationship inference as high as with the initial (large collection) group that we worked on.</p>
<p>Let’s still see what we find for the 50 highest FS/U logl fish in each group.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">final</span> <span class="op">&lt;-</span> <span class="va">next_step</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>matchers <span class="op">=</span> <span class="fu">pmap</span><span class="op">(</span>.l<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GaF</span>, y <span class="op">=</span> <span class="va">CK</span>, z <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>, 
                         .f <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">z</span><span class="op">)</span> <span class="fu"><a href="../reference/find_close_matching_genotypes.html">find_close_matching_genotypes</a></span><span class="op">(</span>LG <span class="op">=</span> <span class="va">x</span><span class="op">$</span><span class="va">Genos</span>, CK <span class="op">=</span> <span class="va">y</span>, max_mismatch <span class="op">=</span> <span class="va">z</span><span class="op">)</span><span class="op">)</span>,
         FS_candidates <span class="op">=</span> <span class="fu">pmap</span><span class="op">(</span>.l <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GaF</span>, y <span class="op">=</span> <span class="va">CK</span><span class="op">)</span>, 
                              .f <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span> <span class="fu"><a href="../reference/pairwise_kin_logl_ratios.html">pairwise_kin_logl_ratios</a></span><span class="op">(</span>D1 <span class="op">=</span> <span class="va">x</span><span class="op">$</span><span class="va">Genos</span>, D2 <span class="op">=</span> <span class="va">x</span><span class="op">$</span><span class="va">Genos</span>, CK <span class="op">=</span> <span class="va">y</span>, numer <span class="op">=</span> <span class="st">"FS"</span>, denom <span class="op">=</span> <span class="st">"U"</span>, keep_top <span class="op">=</span> <span class="fl">20</span>, num_cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">logl_ratio</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">slice</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span><span class="op">)</span>
         <span class="op">)</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span>
<span class="co">#&gt; D1 and D2 are identical: dropping self comparisons and keeping only first instance of each pair</span></code></pre></div>
<p>And now at the very end it would be interesting to plot the highest 50 from each collection. Sounds like a job for unnest.</p>
<p>While we are at it. Let’s figure out what the Logl_ratio value would be that gives us 1/100 * the reciprocal of the number of comparisons.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># first, let's get that lambda_c</span>
<span class="va">DD4</span> <span class="op">&lt;-</span> <span class="va">final</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>sample_size <span class="op">=</span>  <span class="fu">map_int</span><span class="op">(</span>.x <span class="op">=</span> <span class="va">data</span>, .f <span class="op">=</span> <span class="va">nrow</span><span class="op">)</span>,
         coll_str <span class="op">=</span> <span class="fu">str_c</span><span class="op">(</span><span class="va">collection</span>, <span class="st">" ("</span>, <span class="va">sample_size</span>, <span class="st">")"</span><span class="op">)</span>,
         num_pairs <span class="op">=</span> <span class="va">sample_size</span> <span class="op">*</span> <span class="op">(</span><span class="va">sample_size</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span>,
         lambda_crit <span class="op">=</span> <span class="fu">map2_dbl</span><span class="op">(</span>.x <span class="op">=</span> <span class="va">FPRs</span>, .y <span class="op">=</span> <span class="va">num_pairs</span>, 
                                .f <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span> 
                                  <span class="va">x</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">FPR</span> <span class="op">&lt;</span> <span class="fl">0.01</span> <span class="op">/</span> <span class="va">y</span> <span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">slice</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">pull</span><span class="op">(</span><span class="va">Lambda_star</span><span class="op">)</span><span class="op">)</span>
         <span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="va">coll_str</span>, <span class="va">lambda_crit</span>, <span class="va">FS_candidates</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">unnest</span><span class="op">(</span><span class="va">FS_candidates</span><span class="op">)</span>
<span class="co">#&gt; Adding missing grouping variables: `collection`</span>

<span class="co"># now we can plot that nicely</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">DD4</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">coll_str</span>, y <span class="op">=</span> <span class="va">logl_ratio</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"blue"</span>, alpha <span class="op">=</span> <span class="fl">0.2</span>, width <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">lambda_crit</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1.0</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="CKMRsim-example-2-microsatellites_files/figure-html/unnamed-chunk-22-1.png" width="672"></p>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-bradburyGenotypingbysequencingGenomewideMicrosatellite2018">
<p>Bradbury, Ian R., Brendan F. Wringe, Beth Watson, Ian Paterson, John Horne, Robert Beiko, Sarah J. Lehnert, et al. 2018. “Genotyping-by-Sequencing of Genome-Wide Microsatellite Loci Reveals Fine-Scale Harvest Composition in a Coastal Atlantic Salmon Fishery.” <em>Evolutionary Applications</em> 11 (6): 918–30. <a href="https://doi.org/10.1111/eva.12606">https://doi.org/10.1111/eva.12606</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Eric C. Anderson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
