<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>implements the generalized Allele-Based genotyping error model — general_allele_based_geno_err_model • CKMRsim</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="implements the generalized Allele-Based genotyping error model — general_allele_based_geno_err_model" />
<meta property="og:description" content="This is the version into which you pass in your own matrices of allele
call probabilities W (corresponding to omega) and Ws (omega-star)
and vectors of allele-specific dropout rates. Notation follows the paper
for the most part. So see it for details.  Note that this is vectorized
over g, gp, h, and hp.  Note that g, gp, h, and hp can be character vectors
giving the allele names, so long as the vector D has names which are the allele
names and W and Ws have rownames and colnames which are the allele names." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CKMRsim</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/CKMRsim-developer-notes.html">CKMRsim-developer-notes</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-example-1.html">CKMRsim-example-1</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-example-2-microsatellites.html">CKMRsim-example-2-microsatellites</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-pairwise-relationships.html">CKMRsim-pairwise-relationships</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-simulating-linked-markers.html">Simulating-linked-markers</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-writing-geno-error-funcs.html">CKMRsim-writing-geno-error-functions</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>implements the generalized Allele-Based genotyping error model</h1>
    
    <div class="hidden name"><code>general_allele_based_geno_err_model.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This is the version into which you pass in your own matrices of allele
call probabilities W (corresponding to omega) and Ws (omega-star)
and vectors of allele-specific dropout rates. Notation follows the paper
for the most part. So see it for details.  Note that this is vectorized
over g, gp, h, and hp.  Note that g, gp, h, and hp can be character vectors
giving the allele names, so long as the vector D has names which are the allele
names and W and Ws have rownames and colnames which are the allele names.</p>
    </div>

    <pre class="usage"><span class='fu'>general_allele_based_geno_err_model</span><span class='op'>(</span><span class='va'>g</span>, <span class='va'>gp</span>, <span class='va'>h</span>, <span class='va'>hp</span>, <span class='va'>D</span>, <span class='va'>W</span>, <span class='va'>Ws</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>g</th>
      <td><p>allelic type of gene copy g in the true genotype.</p></td>
    </tr>
    <tr>
      <th>gp</th>
      <td><p>allelic type of gene copy g' in the true genotype</p></td>
    </tr>
    <tr>
      <th>h</th>
      <td><p>allelic type of one of the gene copies of the observed genotyped</p></td>
    </tr>
    <tr>
      <th>hp</th>
      <td><p>allelic type of the other gene copy of the observed genotyped</p></td>
    </tr>
    <tr>
      <th>D</th>
      <td><p>vector of allele-specific dropout rates (delta) in the paper. Can be named
with the names of the alleles.</p></td>
    </tr>
    <tr>
      <th>W</th>
      <td><p>the matrix omega of allelic call probabilities. Can have rownames and colnames that
are the allele names.</p></td>
    </tr>
    <tr>
      <th>Ws</th>
      <td><p>the matrix omega* of allelic call probabilities</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>This function returns a vector that is in the order of the genotypes
as they are listed according to  g, gp, h, and hp.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># let's fabricate a locus with 5 alleles named "a1" through "a5"</span>
<span class='co'># and make W random, but with a much higher chance of observing the</span>
<span class='co'># correct allele.</span>
<span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>5</span><span class='op'>)</span>
<span class='va'>alles</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span><span class='op'>(</span><span class='st'>"a"</span>, <span class='fl'>1</span><span class='op'>:</span><span class='fl'>5</span>, sep <span class='op'>=</span> <span class='st'>""</span><span class='op'>)</span>

<span class='co'># here is a data frame of all possible genotypes</span>
<span class='va'>pg</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/expand.grid.html'>expand.grid</a></span><span class='op'>(</span>g<span class='op'>=</span><span class='va'>alles</span>, gp<span class='op'>=</span><span class='va'>alles</span>, stringsAsFactors <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/filter.html'>filter</a></span><span class='op'>(</span><span class='va'>g</span><span class='op'>&lt;=</span><span class='va'>gp</span><span class='op'>)</span>

<span class='co'># and then we replicate that to get all possible combinations of true</span>
<span class='co'># genotypes to observed genotypes</span>
<span class='va'>input_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='va'>pg</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>pg</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>pg</span><span class='op'>)</span><span class='op'>)</span>, <span class='op'>]</span>,
                  <span class='va'>pg</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>pg</span><span class='op'>)</span>, each <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>pg</span><span class='op'>)</span><span class='op'>)</span>, <span class='op'>]</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>input_df</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>3</span><span class='op'>:</span><span class='fl'>4</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"h"</span>, <span class='st'>"hp"</span><span class='op'>)</span>
<span class='va'>input_df</span> <span class='op'>&lt;-</span> <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/tbl_df.html'>tbl_df</a></span><span class='op'>(</span><span class='va'>input_df</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: `tbl_df()` is deprecated as of dplyr 1.0.0.</span>
#&gt; <span class='warning'>Please use `tibble::as_tibble()` instead.</span>
#&gt; <span class='warning'><span style='color: #555555;'>This warning is displayed once every 8 hours.</span><span></span>
#&gt; <span class='warning'></span><span style='color: #555555;'>Call `lifecycle::last_warnings()` to see where this warning was generated.</span><span></span></div><div class='input'>

<span class='co'># here we make the matrix W (just some random numbers)</span>
<span class='va'>W</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/Uniform.html'>runif</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>alles</span><span class='op'>)</span><span class='op'>^</span><span class='fl'>2</span>, <span class='fl'>0.001</span>, <span class='fl'>0.02</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span><span class='op'>(</span>nrow <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>alles</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/dimnames.html'>dimnames</a></span><span class='op'>(</span><span class='va'>W</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='va'>alles</span>, <span class='va'>alles</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/diag.html'>diag</a></span><span class='op'>(</span><span class='va'>W</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/diag.html'>diag</a></span><span class='op'>(</span><span class='va'>W</span><span class='op'>)</span> <span class='op'>+</span> <span class='fl'>1</span>  <span class='co'># make it most likely to call the right allele</span>
<span class='va'>W</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='va'>W</span>, <span class='fl'>1</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='va'>x</span><span class='op'>/</span><span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>  <span class='co'># normalize them to be probablities</span>

<span class='co'># make Wstar to be similar but less likely to miscall</span>
<span class='va'>Ws</span> <span class='op'>&lt;-</span> <span class='va'>W</span>
<span class='fu'><a href='https://rdrr.io/r/base/diag.html'>diag</a></span><span class='op'>(</span><span class='va'>Ws</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fl'>2</span>
<span class='va'>Ws</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='va'>Ws</span>, <span class='fl'>1</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='va'>x</span><span class='op'>/</span><span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># finally make a random D</span>
<span class='va'>D</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/Uniform.html'>runif</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>alles</span><span class='op'>)</span>, min <span class='op'>=</span> <span class='fl'>0.001</span>, max <span class='op'>=</span> <span class='fl'>0.05</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>D</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='va'>alles</span>

<span class='co'># then compute the probs</span>
<span class='va'>gprobs</span> <span class='op'>&lt;-</span> <span class='va'>input_df</span> <span class='op'>%&gt;%</span>
 <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span><span class='op'>(</span>prob <span class='op'>=</span> <span class='fu'>general_allele_based_geno_err_model</span><span class='op'>(</span><span class='va'>g</span>, <span class='va'>gp</span>, <span class='va'>h</span>, <span class='va'>hp</span>, <span class='va'>D</span>, <span class='va'>W</span>, <span class='va'>Ws</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># now, confirm that, for each true genotype, when we sum over the</span>
<span class='co'># probs of the observed genotypes, we get 1.</span>
<span class='va'>gprobs</span> <span class='op'>%&gt;%</span>
 <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span><span class='op'>(</span>ggp <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span><span class='op'>(</span><span class='va'>g</span>, <span class='va'>gp</span>, sep <span class='op'>=</span> <span class='st'>"-"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/group_by.html'>group_by</a></span><span class='op'>(</span><span class='va'>ggp</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/summarise.html'>summarise</a></span><span class='op'>(</span>probSum <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>prob</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>`summarise()` ungrouping output (override with `.groups` argument)</span></div><div class='output co'>#&gt; </span><span style='color: #555555;'># A tibble: 15 x 2</span><span>
#&gt;    ggp   probSum
#&gt;    </span><span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span>   </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>
#&gt; </span><span style='color: #555555;'> 1</span><span> a1-a1    1.  
#&gt; </span><span style='color: #555555;'> 2</span><span> a1-a2    1   
#&gt; </span><span style='color: #555555;'> 3</span><span> a1-a3    1   
#&gt; </span><span style='color: #555555;'> 4</span><span> a1-a4    1.  
#&gt; </span><span style='color: #555555;'> 5</span><span> a1-a5    1   
#&gt; </span><span style='color: #555555;'> 6</span><span> a2-a2    1.00
#&gt; </span><span style='color: #555555;'> 7</span><span> a2-a3    1   
#&gt; </span><span style='color: #555555;'> 8</span><span> a2-a4    1   
#&gt; </span><span style='color: #555555;'> 9</span><span> a2-a5    1   
#&gt; </span><span style='color: #555555;'>10</span><span> a3-a3    1.00
#&gt; </span><span style='color: #555555;'>11</span><span> a3-a4    1.  
#&gt; </span><span style='color: #555555;'>12</span><span> a3-a5    1   
#&gt; </span><span style='color: #555555;'>13</span><span> a4-a4    1.  
#&gt; </span><span style='color: #555555;'>14</span><span> a4-a5    1.  
#&gt; </span><span style='color: #555555;'>15</span><span> a5-a5    1   </div><div class='input'>
<span class='co'># Eureka! it all checks out!</span>
</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Eric C. Anderson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


